test/Microsoft.VisualStudio.Editor.Razor.Test/BraceSmartIndenterTest.cs

```diff
@@ -0,0 +1,382 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.VisualStudio.Editor.Razor.Test.Infrastructure;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Operations;
+using Moq;
+using Xunit;
+
+namespace Microsoft.VisualStudio.Editor.Razor
+{
+    public class BraceSmartIndenterTest : ForegroundDispatcherTestBase
+    {
+        [Fact]
+        public void TextBuffer_OnPostChanged_MovesToEndOfLine()
+        {
+            // Arrange
+            var change = Environment.NewLine;
+            var initialSnapshot = new StringTextSnapshot("@{ }");
+            var afterChangeSnapshot = new StringTextSnapshot("@{ " + change + "}");
+            var edit = new TestEdit(3, 0, initialSnapshot, change.Length, afterChangeSnapshot, change);
+            var expectedIndentResult = "@{ " + change + change + "}";
+
+            var bufferPosition = new VirtualSnapshotPoint(afterChangeSnapshot, 3 + change.Length);
+            var caret = CreateCaretFrom(3 + change.Length, afterChangeSnapshot);
+            TestTextBuffer textBuffer = null;
+            var focusedTextView = CreateFocusedTextView(() => textBuffer, caret);
+            textBuffer = CreateTrackedTextBuffer(initialSnapshot, focusedTextView);
+            var editorOperations = new Mock<IEditorOperations>();
+            editorOperations.Setup(operations => operations.MoveToEndOfLine(false));
+            var editorOperationsFactory = new Mock<IEditorOperationsFactoryService>();
+            editorOperationsFactory.Setup(factory => factory.GetEditorOperations(It.IsAny<ITextView>()))
+                .Returns(editorOperations.Object);
+            var braceSmartIndenter = new BraceSmartIndenter(Dispatcher, textBuffer, editorOperationsFactory.Object);
+
+            // Act
+            textBuffer.ApplyEdit(edit);
+
+            // Assert
+            editorOperations.VerifyAll();
+        }
+
+        [Fact]
+        public void TextBuffer_OnPostChanged_PlacesCursorBackToOriginalLocation()
```

Comment:
`TextBuffer_OnPostChanged_MovesToEndOfLine` and `TextBuffer_OnPostChanged_PlacesCursorBackToOriginalLocation` - what's different about these cases?

Reply:
They're exactly the same just verifying two different aspects of what should happen. In editor land those aspects have impacts on what the user sees. Would rather have 1 test for each so if we break one of those expectations down the line we have 1 specific test failing rather than 1 multi-purpose test.

Reply:
> They're exactly the same just verifying two different aspects of what should happen. 

But why though

> Would rather have 1 test for each so if we break one of those expectations down the line we have 1 specific test failing rather than 1 multi-purpose test.

This just proves that the factoring is wrong. 

These tests (as written) are integration tests. Unit tests have the goal that you're stating, but these are not that.